### PR TITLE
PCGrad fix: regroup ood_cond with in_dist (restore 13.59)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
The current PCGrad (PR #1456) groups ood_cond as OOD because it uses AoA > 1σ as a criterion (line 787: `x[:, 0, 14].abs() > 1.0`). But ood_cond samples have gradients more aligned with in_dist than with tandem/ood_re — the regression from 13.59 to 13.90 confirms this. Removing the AoA criterion from OOD grouping keeps ood_cond samples in Group A (in_dist), restoring the pre-PCGrad ood_cond performance while preserving PCGrad's benefits for tandem and ood_re.

## Instructions
On line 787, change the PCGrad grouping criterion:

```python
# OLD (current):
is_ood_pcgrad = is_tandem_batch | (x[:, 0, 13] > 1.0) | (x[:, 0, 14].abs() > 1.0)
# NEW (fix): only tandem and extreme-Re are OOD group, NOT extreme-AoA
is_ood_pcgrad = is_tandem_batch | (x[:, 0, 13] > 1.0)
```

This removes `(x[:, 0, 14].abs() > 1.0)` so ood_cond samples stay with in_dist in Group A.

Run with `--wandb_group pcgrad-fix-grouping`.

## Baseline
| Split | val_loss | mae_surf_p |
|-------|----------|------------|
| val_in_dist | — | 17.03 |
| val_ood_cond | — | 13.90 (**regressed from 13.59**) |
| val_ood_re | — | 27.62 |
| val_tandem_transfer | — | 38.14 |
| **combined** | **0.8525** | — |

---

### Iteration 1: 2-group fix (remove AoA from OOD)
**W&B run:** ub2qtcau | **Epochs:** 61/100 (timeout) | **Peak memory:** ~17.6 GB

| Split | val_loss | mae_surf_Ux | mae_surf_Uy | mae_surf_p |
|-------|----------|-------------|-------------|------------|
| val_in_dist | 0.5855 | 5.45 | 1.65 | 17.81 |
| val_ood_cond | 0.6754 | 3.34 | 1.11 | **13.54** |
| val_ood_re | 0.5308 | 3.03 | 0.99 | 27.35 |
| val_tandem_transfer | 1.6487 | 5.46 | 2.11 | 39.18 |
| **combined** | **0.8601** | — | — | — |

**mean3 mae_surf_p:** 23.51 | ood_cond recovered 13.90 → 13.54 ✓ but in_dist regressed

### Iteration 2: 3-group PCGrad (A=in_dist+ood_cond, B=tandem, C=ood_re)
**W&B run:** e7y3kx23 | **Epochs:** 51/100 (timeout) | **Peak memory:** ~17.6 GB

| Split | val_loss | mae_surf_Ux | mae_surf_Uy | mae_surf_p |
|-------|----------|-------------|-------------|------------|
| val_in_dist | 0.6189 | 6.88 | 1.49 | 18.80 |
| val_ood_cond | 0.7728 | 4.37 | 1.06 | 15.73 |
| val_ood_re | 0.5744 | 4.03 | 0.89 | 28.30 |
| val_tandem_transfer | 1.6756 | 6.29 | 2.09 | 39.23 |
| **combined** | **0.9104** | — | — | — |

**mean3 mae_surf_p:** 24.59 | **Worse than both 2-group and noam baseline**

### What happened

**3-group approach didn't work.** Three backward passes per step makes each epoch ~19% slower (~37s vs ~31s), limiting training to only 51 epochs vs 61 for 2-group. The ood_re group is a small fraction of most batches, so its gradient estimates are noisy. The extra pairwise projections appear to over-constrain the gradient updates, hurting convergence.

The fundamental problem with the PCGrad family here: the 3 backward passes + pairwise projection computation add significant overhead without a clear benefit within the 30-minute training budget.

### Summary comparison
| Config | val/loss | mean3 | ood_cond surf_p |
|--------|----------|-------|-----------------|
| noam baseline | 0.8555 | ~23.20 | ~13.59 |
| 2-group (baseline for this PR) | 0.8525 | 23.02 | 13.90 |
| 2-group fix (iter 1) | 0.8601 | 23.51 | 13.54 |
| **3-group (iter 2)** | **0.9104** | **24.59** | **15.73** |

PCGrad does not appear to be beneficial for this model/dataset combination.

### Suggested follow-ups

- PCGrad is probably not the right lever here. The adaptive_boost mechanism already handles tandem/in_dist imbalance. May be worth confirming this with an ablation.
- If gradient conflict is genuinely the bottleneck, a much lighter approach would be to just scale the learning rate differently per group (via the tandem_boost), which is simpler and doesn't require extra backward passes.
- Explore other directions entirely: architecture changes, better positional encoding, or augmentation.

### CODE REVIEW

**Comment by tcapelle** [comment]
## Review: Promising direction, needs refinement

**Results confirmed via W&B:** val_loss=0.8601 (vs baseline 0.8525). ood_cond restored from 13.90 → 13.54 — the grouping diagnosis was correct! But in_dist regressed (17.03→17.81) and tandem got worse (38.14→39.18).

**Root cause:** With only tandem + extreme-Re as OOD, the OOD group is too small/imbalanced relative to the in_dist+ood_cond group. The gradient projection favors the larger group.

**Next iteration: PCGrad with 3 groups instead of 2.**

The key insight from this round (confirmed across your PR and gilbert's ablation #1460): ood_cond is aligned with in_dist, but tandem and ood_re have different gradient directions. A 3-group PCGrad would handle this:

- **Group A**: in_dist + ood_cond (aligned, share gradients naturally)
- **Group B**: tandem samples only
- **Group C**: extreme-Re samples only (ood_re)

For 3 groups, modify the projection logic:
```python
# Define 3 groups
is_tandem_group = is_tandem_batch
is_ood_re_group = (~is_tandem_batch) & (x[:, 0, 13] > 1.0)
is_indist_group = ~is_tandem_group & ~is_ood_re_group  # includes ood_cond

# Only do PCGrad if at least 2 groups are present
groups_present = [is_indist_group.any(), is_tandem_group.any(), is_ood_re_group.any()]
n_groups = sum(groups_present)

if n_groups >= 2:
    # Compute per-group losses and gradients
    # For each pair of groups: if dot product < 0, project out conflicting component
    # Average all (potentially projected) gradients
```

The 3-group projection: for each pair (A,B), (A,C), (B,C), check if gradients conflict. If so, project out the conflicting component. Then average all projected gradients. This is the standard multi-task PCGrad extension.

Run with `--wandb_group pcgrad-3group`.
